File: core/IdempotencyService.php
<?php
/**
 * OmniDesk AI — Idempotency Service (Phase 14)
 *
 * Enforces transaction-level idempotency for high-risk write operations.
 * Prevents double-charging, duplicate invoices, and double-submits.
 */

namespace Core;

use Exception;

class IdempotencyService
{
    /**
     * Check if a request has already been processed for the workspace.
     */
    public static function check(string $key, int $workspaceId, int $userId, string $toolName, array $payload): ?array
    {
        if (empty($key)) {
            return null;
        }

        $requestHash = hash('sha256', json_encode($payload));

        try {
            $db = Database::getInstance();
            $record = $db->fetchOne(
                'SELECT * FROM idempotency_keys 
                 WHERE idempotency_key = :key AND workspace_id = :ws LIMIT 1',
                ['key' => $key, 'ws' => $workspaceId]
            );

            if ($record) {
                return [
                    'is_duplicate'     => true,
                    'status'           => $record['status'],
                    'response_payload' => json_decode($record['response_payload'] ?? '{}', true),
                    'transaction_id'   => $record['transaction_id'],
                    'created_at'       => $record['created_at']
                ];
            }
        } catch (\Throwable $e) {
            // Fallback for environment without DB
        }

        return ['is_duplicate' => false, 'request_hash' => $requestHash];
    }

    /**
     * Store the completed transaction response against the idempotency key.
     */
    public static function record(string $key, int $workspaceId, int $userId, string $toolName, array $payload, array $response, ?int $transactionId = null): bool
    {
        if (empty($key)) {
            return false;
        }

        $requestHash = hash('sha256', json_encode($payload));

        try {
            $db = Database::getInstance();
            $db->execute(
                'INSERT INTO idempotency_keys (idempotency_key, workspace_id, user_id, tool_name, request_hash, response_payload, transaction_id, status, created_at)
                 VALUES (:key, :ws, :uid, :tool, :rhash, :resp, :txid, "completed", NOW())
                 ON DUPLICATE KEY UPDATE response_payload = :resp2, status = "completed", updated_at = NOW()',
                [
                    'key'   => $key,
                    'ws'    => $workspaceId,
                    'uid'   => $userId,
                    'tool'  => $toolName,
                    'rhash' => $requestHash,
                    'resp'  => json_encode($response),
                    'resp2' => json_encode($response),
                    'txid'  => $transactionId
                ]
            );
            return true;
        } catch (\Throwable $e) {
            return false;
        }
    }
}

File: ai/app/services/idempotency_service.py
"""
OmniDesk AI — Transaction Idempotency Service (Phase 14)

Provides single-execution guarantees for high-risk write operations.
"""

from typing import Dict, Any, Optional
import hashlib
import json
from datetime import datetime

class IdempotencyService:
    # Workspace-scoped idempotency store: {ws_id: {idempotency_key: record}}
    _store: Dict[int, Dict[str, Dict[str, Any]]] = {}

    @classmethod
    def check(cls, ws_id: int, key: str, payload: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        if not key:
            return None

        ws_store = cls._store.get(ws_id, {})
        record = ws_store.get(key)
        if record:
            return {
                "is_duplicate": True,
                "status": record["status"],
                "response_payload": record["response_payload"],
                "created_at": record["created_at"]
            }
        return {"is_duplicate": False}

    @classmethod
    def record(cls, ws_id: int, user_id: int, tool_name: str, key: str, payload: Dict[str, Any], response: Dict[str, Any]):
        if not key:
            return

        if ws_id not in cls._store:
            cls._store[ws_id] = {}

        cls._store[ws_id][key] = {
            "user_id": user_id,
            "tool_name": tool_name,
            "request_hash": hashlib.sha256(json.dumps(payload, sort_keys=True).encode('utf-8')).hexdigest(),
            "response_payload": response,
            "status": "completed",
            "created_at": datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        }
